source-dropbox/main.go

```diff
@@ -0,0 +1,312 @@
+	Credentials *Credentials   `json:"credentials"`
+	Parser      *parser.Config `json:"parser"`
+	MatchKeys   string         `json:"matchKeys,omitempty"`
+	Advanced    advancedConfig `json:"advanced"`
+}
+
+type advancedConfig struct {
+	AscendingKeys bool `json:"ascendingKeys,omitempty"`
+}
+
+func (c config) Validate() error {
+	requiredProperties := [][]string{
+		{"Path", c.Path},
+	}
+	for _, req := range requiredProperties {
+		if req[1] == "" {
+			return fmt.Errorf("missing '%s'", req[0])
+		}
+	}
+	return nil
+}
+
+func (c config) DiscoverRoot() string {
+	return c.Path
+}
+
+func (c config) RecommendedName() string {
+	return c.Path[1:]
+}
+
+func (c config) FilesAreMonotonic() bool {
+	return c.Advanced.AscendingKeys
+}
+
+func (c config) ParserConfig() *parser.Config {
+	return c.Parser
+}
+
+func (c config) PathRegex() string {
+	return c.MatchKeys
+}
+
+func newDropboxStore(ctx context.Context, cfg config) (*dropboxStore, error) {
+	log.Debug("Creating Dropbox client")
+	oauthClient, err := cfg.Credentials.GetOauthClient(ctx)
+	if err != nil {
+		return &dropboxStore{}, err
+	}
+	config := dropbox.Config{
+		//! Note: This Client property was originally build for testing purposes. So, it can change in the future.
+		Client:   oauthClient,
+		LogLevel: dropbox.LogDebug,
+	}
+	client := files.New(config)
+
+	store := dropboxStore{
+		client: client,
+		config: cfg,
+	}
+	if err := store.check(); err != nil {
+		return &dropboxStore{}, err
+	}
+	return &store, nil
+}
+
+type dropboxStore struct {
+	client files.Client
+	config config
+}
+
+// check checks the connection to the Azure Blob Storage container.
+// It returns an error if the container is not found, the account is disabled, or if there is an authorization failure.
+// If the listing is successful, it returns nil.
+func (dbx *dropboxStore) check() error {
+	resp, err := dbx.client.ListFolder(&files.ListFolderArg{Path: dbx.config.Path})
+	if err != nil {
+		log.Error("Failed to list files")
+		return err
+	}
+	for _, entry := range resp.Entries {
+		fileEntry, ok := entry.(*files.FileMetadata)
+		if !ok {
+			continue
+		}
+		meta, reader, err := dbx.client.Download(&files.DownloadArg{Path: fileEntry.PathDisplay})
+		if err != nil {
+			log.Error("Failed to download file: ", fileEntry.PathDisplay)
+			return err
+		}
+		log.Debug("File metadata: ", meta)
+		defer reader.Close()
+		content := make([]byte, meta.Size)
+		if _, err := reader.Read(content); err != nil {
+			log.Error("Failed to download file: ", fileEntry.PathDisplay)
+			return err
+		}
+
+		// log.Debug("File content: ", string(content))
+		break
+	}
+	log.Debug("Successfully connected to Dropbox")
+	return nil
+}
+
+func (dbx *dropboxStore) List(ctx context.Context, query filesource.Query) (filesource.Listing, error) {
+	log.Debug("Listing files in path: ", dbx.config.Path)
+	log.Debug("Query: ", query)
+	files, err := dbx.client.ListFolder(&files.ListFolderArg{Path: query.Prefix, Recursive: query.Recursive})
```

Comment:
The `files.ListFolderResult` returned here has a `hasMore` property to indicate that more paged results are available, and there's a `ListFolderContinue` on the `Client` interface that seems to be for the purposes of getting the next page of results. Do we need to be using these capabilities to make sure we get the full listing? It seems like we probably do.